lib/propolis/src/block/crucible.rs

```diff
@@ -11,9 +11,62 @@ use crate::vmm::MemCtx;
 use crucible::{BlockIO, Buffer, CrucibleError, SnapshotDetails, Volume};
 use crucible_client_types::VolumeConstructionRequest;
 use oximeter::types::ProducerRegistry;
+use slog::{error, info, Logger};
 use thiserror::Error;
 use uuid::Uuid;
 
+use internal_dns::resolver::{ResolveError, Resolver};
+use internal_dns::ServiceName;
+pub use nexus_client::Client as NexusClient;
+use omicron_common::address::NEXUS_INTERNAL_PORT;
+use std::net::Ipv6Addr;
+use std::net::SocketAddr;
+
+// For communicating to nexus once the scrubber is done.
+struct Inner {
+    log: Logger,
+    resolver: Resolver,
+}
+
+/// Wrapper around a [`NexusClient`] object, which allows deferring
+/// the DNS lookup until accessed.
+///
+/// Without the assistance of OS-level DNS lookups, the [`NexusClient`]
+/// interface requires knowledge of the target service IP address.
+/// For some services, like Nexus, this can be painful, as the IP address
+/// may not have even been allocated when the Sled Agent starts.
+///
+/// This structure allows clients to access the client on-demand, performing
+/// the DNS lookup only once it is actually needed.
+#[derive(Clone)]
+pub struct LazyNexusClient {
+    inner: Arc<Inner>,
+}
+
+impl LazyNexusClient {
+    pub fn new(log: Logger, addr: Ipv6Addr) -> Result<Self, ResolveError> {
+        Ok(Self {
+            inner: Arc::new(Inner {
+                log: log.clone(),
+                resolver: Resolver::new_from_ip(log, addr)?,
+            }),
+        })
+    }
+
+    pub async fn get_ip(&self) -> Result<Ipv6Addr, ResolveError> {
+        self.inner.resolver.lookup_ipv6(ServiceName::Nexus).await
+    }
+
+    pub async fn get(&self) -> Result<NexusClient, ResolveError> {
+        let address = self.get_ip().await?;
+
+        Ok(NexusClient::new(
+            &format!("http://[{}]:{}", address, NEXUS_INTERNAL_PORT),
```

Comment:
I would use `lookup_socket_v6` instead of `lookup_ipv6` so you don't have to hardcode the port here.

Reply:
Nice, and done.